streaming/base/dataset/streaming.py

```diff
@@ -0,0 +1,638 @@
+        self.split = split or ''  # Empty string for os.path.join().
+        self.shuffle = shuffle
+        self.predownload = predownload
+        self.keep_zip = keep_zip
+        self.download_retry = download_retry
+        self.download_timeout = download_timeout
+        self.validate_hash = validate_hash or None
+        # Seed is set below.
+        self.batch_size = batch_size
+        self.num_workers = num_workers
+
+        # Load the index.json file.
+        basename = get_index_basename()
+        world = World()
+        if world.is_local_leader:
+            filename = self._download_file(basename)
+        else:
+            filename = os.path.join(local, split, basename)  # pyright: ignore
+        dist.barrier()
+        obj = json.load(open(filename))
+        assert obj['version'] == 2
+
+        # Initialize shard readers according to the loaded info.
+        self.shards = []
+        for info in obj['shards']:
+            shard = reader_from_json(local, split, info)
+            self.shards.append(shard)
+
+        # Build the Index (for partitioning and mapping samples to shards).
+        self.shard_sizes = np.array([x.samples for x in self.shards])
+        self.index = Index(self.shard_sizes)
+
+        # Setup for coordinating.
+        device = torch.device(f'cuda:{world.rank_of_node}')
+        tensor = torch.zeros(1, dtype=torch.int64, device=device)
+
+        # Coordinate the seed across ranks.
+        if world.is_leader:
+            if seed is None:
+                seed = np.random.randint(1 << 60)
+            tensor[0] = seed
+        dist.broadcast(tensor, 0)
+        self.seed = int(tensor)
+
+        # Add a coordinated random prefix to all shm names for uniqueness.
+        if world.is_leader:
+            tensor[0] = np.random.randint(1 << 60)
+        dist.broadcast(tensor, 0)
+        self._prefix = f'{int(tensor):016x}_{self.split}'
+
+        # Set up the epoch counter.
+        #
+        # Note: we do not assume that the end of __iter__() will ever be reached, so we need to
+        # increment the epoch counter at the start of __iter__() instead of at the end, so we need
+        # to track what the next epoch is, not the current epoch.
+        name = f'{self._prefix}_next_epoch'
+        size = np.int64().nbytes
+        try:
+            self._next_epoch_shm = SharedMemory(name, True, size)
+        except FileExistsError:
+            self._next_epoch_shm = SharedMemory(name)
+        self._next_epoch_arr = np.ndarray(1, buffer=self._next_epoch_shm.buf, dtype=np.int64)
+        self._next_epoch_arr[0] = 0
+
+        # Placeholder for _resume_shm, a shared memory object where load_state_dict() saves its
+        # data to be picked up by __iter__().
+        self._resume_shm = None
+
+        # Create the barrier.
+        total_workers = world.ranks_per_node * (self.num_workers or 1)
+        self._barrier_filelock_path = os.path.join('/tmp', 'mds', self._prefix, 'barrier_filelock')
```

Comment:
Few things:
1. use `os.sep` instead of `/`.
2. why `mds` ? since we support multiple writers of different type. Can we get `mds` name during runtime ?

Reply:
Fixed